File: BINDINGS/cxx/eflxx_examples/src/elementaryxx/full/test_progressbar.cpp
#include "test.h"

#include <ecorexx/Ecorexx.h>

typedef struct _MyProgressbar
{
  Progressbar *pb1;
  Progressbar *pb2;
  Progressbar *pb3;
  Progressbar *pb4;
  Progressbar *pb5;
  Progressbar *pb6;
  Progressbar *pb7;
  bool run;
  Ecorexx::Timer *timer;
} MyProgressbar;

static MyProgressbar _test_progressbar;

bool _my_progressbar_value_set (Ecorexx::Timer &timer)
{
  double progress;

  progress = _test_progressbar.pb1->getValue ();

  if (progress < 1.0)
  {
    progress += 0.0123;
  }
  else
  {
    progress = 0.0;
  }

  _test_progressbar.pb1->setValue (progress);
  _test_progressbar.pb4->setValue (progress);
  _test_progressbar.pb3->setValue (progress);
  _test_progressbar.pb6->setValue (progress);

  if (progress > 1.0)
  {
    _test_progressbar.run = false;
    return false; // don't run again and delete timer
  }

  return true;
}

static void my_progressbar_test_start (Evasxx::Object &obj, void *event_info)
{
  _test_progressbar.pb2->pulse (true);
  _test_progressbar.pb5->pulse (true);
  _test_progressbar.pb7->pulse (true);

  if (!_test_progressbar.run)
  {
    sigc::slot <bool, Ecorexx::Timer&> timerSlot = sigc::ptr_fun (&_my_progressbar_value_set);

    _test_progressbar.timer = Ecorexx::Timer::factory (0.1, timerSlot);

    _test_progressbar.run = true;
  }
}

static void _test_stop ()
{
  _test_progressbar.pb2->pulse (false);
  _test_progressbar.pb5->pulse (false);
  _test_progressbar.pb7->pulse (false);

  if (_test_progressbar.run)
  {
     _test_progressbar.timer->destroy ();
     _test_progressbar.run = false;
  }
}

static void my_progressbar_test_stop (Evasxx::Object &obj, void *event_info)
{
  _test_stop ();
}

static void my_progressbar_destroy (Evasxx::Object &obj, void *event_info)
{
  _test_stop ();
  delete &obj;
}

void test_progressbar(void *data, Evas_Object *obj, void *event_info)
{
  Progressbar *pb = NULL;
  Icon *ic1 = NULL;
  Icon *ic2 = NULL;
  Button *bt = NULL;

  // start in stop mode
  _test_progressbar.run = false;

  Window *win = Window::factory ("progressbar", ELM_WIN_BASIC);
  win->setTitle ("Progressbar");
  win->getEventSignal ("delete,request")->connect (sigc::ptr_fun (&my_progressbar_destroy));

  Background *bg = Background::factory (*win);
  win->addObjectResize (*bg);
  bg->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  bg->show ();

  Box *bx = Box::factory (*win);
  win->addObjectResize (*bx);
  bx->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  bx->show ();

  pb = Progressbar::factory (*win);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  bx->packEnd (*pb);
  pb->show ();
  _test_progressbar.pb1 = pb;

  pb = Progressbar::factory (*win);
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  pb->setLabel ("Infinite bounce");
  pb->setPulse (true);
  bx->packEnd (*pb);
  pb->show ();
  _test_progressbar.pb2 = pb;

  ic1 = Icon::factory (*win);
  ic1->setFile (searchPixmapFile ("elementaryxx/logo_small.png"));
  ic1->setAspectHintSize (EVAS_ASPECT_CONTROL_VERTICAL, Size (1, 1));

  pb = Progressbar::factory (*win);
  pb->setLabel ("Label");
  pb->setIcon (*ic1);
  pb->setInverted (true);
  pb->setUnitFormat ("%1.1f units");
  pb->setSpanSize (200);
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  bx->packEnd (*pb);
  ic1->show ();
  pb->show ();
  _test_progressbar.pb3 = pb;

  Box *hbx = Box::factory (*win);
  hbx->setOrientation (Box::Horizontal);
  hbx->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  hbx->setAlignHintSize (EVAS_HINT_FILL, EVAS_HINT_FILL);
  bx->packEnd (*hbx);
  hbx->show ();

  pb = Progressbar::factory (*win);
  pb->setOrientation (Progressbar::Vertical);
  pb->setAlignHintSize (EVAS_HINT_FILL, EVAS_HINT_FILL);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  hbx->packEnd (*pb);
  pb->setSpanSize (60);
  pb->setLabel ("percent");
  pb->show ();
  _test_progressbar.pb4 = pb;

  pb = Progressbar::factory (*win);
  pb->setOrientation (Progressbar::Vertical);
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  pb->setSpanSize (80);
  pb->hideUnitFormat ();
  pb->setLabel ("Infinite bounce");
  hbx->packEnd (*pb);
  pb->show ();
  _test_progressbar.pb5 = pb;

  ic2 = Icon::factory (*win);
  ic2->setFile (searchPixmapFile ("elementaryxx/logo_small.png"));
  ic2->setAspectHintSize (EVAS_ASPECT_CONTROL_HORIZONTAL, Size (1, 1));

  pb = Progressbar::factory (*win);
  pb->setOrientation (Progressbar::Vertical);
  pb->setLabel ("Label");
  pb->setIcon (*ic2);
  pb->setInverted (true);
  pb->setUnitFormat ("%1.2f%%");
  pb->setSpanSize (200);
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  hbx->packEnd (*pb);
  pb->show ();
  _test_progressbar.pb6 = pb;

  pb = Progressbar::factory (*win);
  pb->setStyle ("wheel");
  pb->setLabel ("Style: wheel");
  pb->setAlignHintSize (EVAS_HINT_FILL, 0.5);
  pb->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  bx->packEnd (*pb);
  pb->show ();
  _test_progressbar.pb7 = pb;

  Box *bt_bx = Box::factory (*win);
  bt_bx->setOrientation (Box::Horizontal);
  bt_bx->setWeightHintSize (EVAS_HINT_EXPAND, EVAS_HINT_EXPAND);
  bx->packEnd (*bt_bx);
  bt_bx->show ();

  bt = Button::factory (*win);
  bt->setLabel ("Start");
  bt->getEventSignal ("clicked")->connect (sigc::ptr_fun (&my_progressbar_test_start));
  bt_bx->packEnd (*bt);
  bt->show ();

  bt = Button::factory (*win);
  bt->setLabel ("Stop");
  bt->getEventSignal ("clicked")->connect (sigc::ptr_fun (&my_progressbar_test_stop));
  bt_bx->packEnd (*bt);
  bt->show ();

  win->show ();
}

File: BINDINGS/python/python-elementary/tests/test_progressbar.py
#!/usr/bin/env python
import os
import elementary
import edje
import ecore
import evas

# ----- Progressbar {{{
my_progressbar_run = False
my_progressbar_timer = None

def my_progressbar_value_set(pb1, pb2, pb3, pb4, pb5, pb6, pb7):
    progress = pb1.value_get()
    if progress < 1.0:
        progress += 0.0123
    else:
        progress = 0.0
    pb1.value_set(progress)
    pb4.value_set(progress)
    pb3.value_set(progress)
    pb6.value_set(progress)
    if progress < 1.0:
        return ecore.ECORE_CALLBACK_RENEW
    global my_progressbar_run
    my_progressbar_run = False
    return ecore.ECORE_CALLBACK_CANCEL

def my_progressbar_test_start(obj, *args, **kwargs):
    (pb1, pb2, pb3, pb4, pb5, pb6, pb7) = args
    pb2.pulse(True)
    pb5.pulse(True)
    pb7.pulse(True)
    global my_progressbar_run
    global my_progressbar_timer
    if not my_progressbar_run:
        my_progressbar_timer = ecore.timer_add(0.1, my_progressbar_value_set,
                                               *args)
        my_progressbar_run = True

def my_progressbar_test_stop(obj, pb1, pb2, pb3, pb4, pb5, pb6, pb7):
    pb2.pulse(False)
    pb5.pulse(False)
    pb7.pulse(False)
    global my_progressbar_run
    global my_progressbar_timer
    if my_progressbar_run:
        my_progressbar_timer.delete()
        my_progressbar_run = False

def my_progressbar_destroy(obj, *args):
    my_progressbar_test_stop(None, *args)
    obj.delete()

def progressbar_clicked(obj, it):
    win = elementary.Window("progressbar", elementary.ELM_WIN_BASIC)
    win.title_set("Progressbar test")

    bg = elementary.Background(win)
    win.resize_object_add(bg)
    bg.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    bg.show()

    bx = elementary.Box(win)
    win.resize_object_add(bx)
    bx.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    bx.show()

    pb1 = elementary.Progressbar(win)
    pb1.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    pb1.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    bx.pack_end(pb1)
    pb1.show()

    pb2 = elementary.Progressbar(win)
    pb2.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    pb2.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    pb2.label_set("Infinite bounce")
    pb2.pulse_set(True)
    bx.pack_end(pb2)
    pb2.show()

    ic1 = elementary.Icon(win)
    ic1.file_set('images/logo_small.png')
    ic1.size_hint_aspect_set(evas.EVAS_ASPECT_CONTROL_VERTICAL, 1, 1)

    pb3 = elementary.Progressbar(win)
    pb3.label_set("Label")
    pb3.icon_set(ic1)
    pb3.inverted_set(True)
    pb3.unit_format_set("%1.1f units")
    pb3.span_size_set(200)
    pb3.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    pb3.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    bx.pack_end(pb3)
    ic1.show()
    pb3.show()

    hbx = elementary.Box(win)
    hbx.horizontal_set(True)
    hbx.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    hbx.size_hint_align_set(evas.EVAS_HINT_FILL, evas.EVAS_HINT_FILL)
    bx.pack_end(hbx)
    hbx.show()

    pb4 = elementary.Progressbar(win)
    pb4.horizontal_set(False)
    pb4.size_hint_align_set(evas.EVAS_HINT_FILL, evas.EVAS_HINT_FILL)
    pb4.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    hbx.pack_end(pb4)
    pb4.span_size_set(60)
    pb4.label_set("percent")
    pb4.show()

    pb5 = elementary.Progressbar(win)
    pb5.horizontal_set(False)
    pb5.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    pb5.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    pb5.span_size_set(80)
    pb5.pulse_set(True)
    pb5.unit_format_set(None)
    pb5.label_set("Infinite bounce")
    hbx.pack_end(pb5)
    pb5.show()

    ic2 = elementary.Icon(win)
    ic2.file_set('images/logo_small.png')
    ic2.size_hint_aspect_set(evas.EVAS_ASPECT_CONTROL_HORIZONTAL, 1, 1)

    pb6 = elementary.Progressbar(win)
    pb6.horizontal_set(False)
    pb6.label_set("Label")
    pb6.icon_set(ic2)
    pb6.inverted_set(True)
    pb6.unit_format_set("%1.2f%%")
    pb6.span_size_set(200)
    pb6.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    pb6.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    hbx.pack_end(pb6)
    ic2.show()
    pb6.show()

    pb7 = elementary.Progressbar(win)
    pb7.style_set("wheel")
    pb7.label_set("Style: wheel")
    pb7.size_hint_align_set(evas.EVAS_HINT_FILL, 0.5)
    pb7.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    bx.pack_end(pb7)
    pb7.show()

    bt_bx = elementary.Box(win)
    bt_bx.horizontal_set(True)
    bt_bx.size_hint_weight_set(evas.EVAS_HINT_EXPAND, evas.EVAS_HINT_EXPAND)
    bx.pack_end(bt_bx)
    bt_bx.show()

    pbt = (pb1, pb2, pb3, pb4, pb5, pb6, pb7)

    bt = elementary.Button(win)
    bt.label_set("Start")
    bt.callback_clicked_add(my_progressbar_test_start, *pbt)
    bt_bx.pack_end(bt)
    bt.show()

    bt = elementary.Button(win)
    bt.label_set("Stop")
    bt.callback_clicked_add(my_progressbar_test_stop, *pbt)
    bt_bx.pack_end(bt)
    bt.show()

    win.callback_destroy_add(my_progressbar_destroy, *pbt)
    win.show()
# }}}

#----- Main -{{{-
if __name__ == "__main__":
    elementary.init()

    progressbar_clicked(None, None)

    elementary.run()
    elementary.shutdown()
# }}}
# vim:foldmethod=marker
